behavioral/observer/observer_module.f90

```diff
@@ -0,0 +1,159 @@
+!> Reference: https://refactoring.guru/design-patterns/observer/go/example
+module observer_pattern
+
+    implicit none
+    private
+
+    public :: item, customer, new_item
+
+    !> Abstract classes
+    type, abstract :: subject
+    contains
+        procedure(register_procedure), deferred :: register
+        procedure(deregister_procedure), deferred :: deregister
+        procedure(notify_all_procedure), deferred :: notify_all
+    end type subject
+
+    type, abstract :: observer
+    contains
+        procedure(update_procedure), deferred :: update
+        procedure(get_ID_procedure), deferred :: get_ID
+    end type observer
+    
+    !> We cannot directly use `class(observer), allocatable :: o_list(:)` 
+    !>  instead of `type(node), allocatable :: o_list(:)`.
+    type node
+        class(observer), allocatable :: o
+    end type node
+
+    abstract interface
+        subroutine register_procedure(self, o)
+            import subject, observer
+            class(subject), intent(inout) :: self
+            class(observer), intent(inout) :: o
+        end subroutine register_procedure
+        subroutine deregister_procedure(self, o)
+            import subject, observer
+            class(subject), intent(inout) :: self
+            class(observer), intent(inout) :: o
+        end subroutine deregister_procedure
+        subroutine notify_all_procedure(self)
+            import subject
+            class(subject), intent(inout) :: self
+        end subroutine notify_all_procedure
+        subroutine update_procedure(self, s)
+            import observer
+            class(observer), intent(inout) :: self
+            character(len=*), intent(inout) :: s 
+        end subroutine update_procedure
+        function get_ID_procedure(self) result(result)
+            import observer
+            class(observer), intent(inout) :: self
+            character(len=:), allocatable :: result
+        end function get_ID_procedure
+    end interface
+
+    !> Specific objects
+
+    type, extends(subject) :: item
+        type(node), allocatable :: o_list(:)
```

Comment:
I want to use `class(observer), allocatable :: o_list(:)` here, but this writing method faces some incompatibility issues with the existing grammar of Fortran. I need help! Who can try to solve this problem?

Maybe here my code is breaking the Liskov substitution principle: The functions should accept arguments from the base classes.